### PR TITLE
aruha-190 define offsets format

### PR DESCRIPTION
**Pull request by rcillo**
Up to now Nakadi Cursors were advertised to be opaque entities and
clients should not make assumptions about its structure and meaning.

But we've been observing clients making assumptions and coding based on
them. This could lead to a major confusion in the future if we no longer
hold those assumptions true.

In order to mitigate this risk, we are committing to a well defined
structure beforehand. It's inspired by Postgres WAL format
https://www.postgresql.org/docs/9.2/static/wal-internals.html and we aim
to address two issues with this format:

1. Make Nakadi operations easier (this feature would have been
tremendously useful for cluster migration or implementing partition
rebalacing)
2. Provide a well defined set of guarantees for consumers to rely one.

**update**

As for why 27 characters = 9 (max integer 7FFFFFFF) + separator + 17 (max long 7FFFFFFFFFFFFFFF).

### CODE REVIEW

File: api/nakadi-event-bus-api.yaml [hunk 1]
```diff
@@ -1184,8 +1184,17 @@ definitions:
...
+
+          - offsets have a fixed size of 26 characters ranging from 0 to F (Hexadecimals).
+          - the first 9 characters are used to define the timeline, a monotonically increasing sequence. Starts with
+            000000001.
+          - the remaining 17 characters defines the index, a strictly increasing sequence in the scope of a timeline.
+            This maps Kafka offsets just by converting to hexadecimals.
+          - offsets are fully ordered and comparable lexicographicaly.
+          - it's not possible to determine how many events there are between two offsets from different timelines.
+          - timeline and index are always padded by zeros.
+          - the first event in a log has a special offset of BEGIN.
```
Comment (v-stepanov) [inline api/nakadi-event-bus-api.yaml]: why? I think the first event has offset zero

Comment (v-stepanov) [inline api/nakadi-event-bus-api.yaml]: Virtual event before the first one has offset BEGIN

**Comment by rcillo** [comment]
@valgog @dehora @antban @v-stepanov @v1ctor @adyach what are your thoughts on this?

I would say that it's going to be hard to implement. 

1. in the code there are many direct casts from String to Long which would be broken.
2. not been able to count how many events there are between two cursors from different timelines would break the functionality from SubscriptionStats

On the other hand, we could just trim the timeline part and keep the implementation as is, delaying the cost of a real implementation.

**Comment by v-stepanov** [comment]
Don't we want to separate the parts by separator, as we advertise cursor as a string anyway ?

**Comment by rcillo** [comment]
@v-stepanov one could argue that a separator would be redundant since the sizes are fixed and defined. On the other hand, it could make implementation less error prone, since splitting by a separator is more guaranteed than splitting by index.

**Comment by v-stepanov** [comment]
@rcillo I think as those are two separate logical parts It doesn't make much sense to merge it. I would choose the implementation with separator. But let's see what others say.

**Comment by rcillo** [comment]
I've added a separator and removed the possibility for doing arithmetics.

File: api/nakadi-event-bus-api.yaml [hunk 2]
```diff
@@ -1186,15 +1186,17 @@ definitions:
         description: |
           Offset of the event being pointed to. Regarding its format and guarantees:
 
-          - offsets have a fixed size of 26 characters ranging from 0 to F (Hexadecimals).
+          - offsets have a fixed size of 27 characters ranging from 0 to F (Hexadecimals).
```
Comment (valgog) [inline api/nakadi-event-bus-api.yaml]: Formally speaking after introduction of a separator `/` these 27 characters are not ranging from 0 to F. One of them is `/`

Comment (valgog) [inline api/nakadi-event-bus-api.yaml]: maybe a regex would be an easier as a documentation? 

`^([0-F]{9})/([0-F]{17})$`

File: api/nakadi-event-bus-api.yaml [hunk 3]
```diff
@@ -1184,8 +1184,19 @@ definitions:
       offset:
```
Comment (adyach) [inline api/nakadi-event-bus-api.yaml]: @rcillo I think it is good to see an example here

File: api/nakadi-event-bus-api.yaml (hunk 2 above)
Comment (rcillo) [inline api/nakadi-event-bus-api.yaml]: Fixed in https://github.com/zalando/nakadi/pull/403/commits/de05a98ad2fe2fa3323cb2b6e0e39afc00362ff7

I think that after adding an example it became more clear. I would say that the regex could just make it more scary to some people.

File: api/nakadi-event-bus-api.yaml (hunk 3 above)
Comment (rcillo) [inline api/nakadi-event-bus-api.yaml]: Done 👍 https://github.com/zalando/nakadi/pull/403/commits/663d602655f23f56a97a86296a009dcec4fd3bf2

**Comment by tlossen** [comment]
hmmmm ... i see the convenience of having offsets scoped by timeline. how about introducing timelines in a more backwards-compatible way?

a long is 64 bit, which we could divide into timeline (16 bit) plus offset (48 bit). in hex, that would be 16 characters, for example 000100000000babe (timeline 0001, offset 00000000babe).

this way, Long.parseLong(value, 16) would still give me a usable long offset that can be compared arithmetically, not only lexicographically.

48 bit is plenty! i did a quick calculation, it would last 8500 years if you publish 1000 events per second. 65K timelines should also be enough for now.

**Comment by rcillo** [comment]
@tlossen we've seem some clients doing arithmetics with offsets. Something like `number_of_events_to_be_consumed = latest_offset_available - current_consumed_offset`. If we make it "backward compatible" (I quoted it because such guarantess have never been in the contract), it would just be delaying errors that should be spotted as soon as possible.

One of the motivations for this change is to push clients to fix their implementations and stop making dangerous assumptions over offsets.

**Comment by tkrop** [comment]
Do you really want to expose Kafka implementation details in this kind. Where is the sugar Nakadi may be able to add to Kafka in the future, if you "fix" it this way?

**Comment by rcillo** [comment]
@tkrop could you please be more specific?

File: api/nakadi-event-bus-api.yaml [hunk 4]
```diff
@@ -1184,8 +1184,20 @@ definitions:
...
-          Offset of the event being pointed to.
-
+          Offset of the event being pointed to. Regarding its format and guarantees:
+
+          - offsets have a fixed size of 27 characters.
+          - the first 9 characters are used to define the timeline, a monotonically increasing sequence. Starts with
+            000000001.
+          - its followed by the separator "/".
+          - the remaining 17 characters defines the index, a strictly increasing sequence in the scope of a timeline.
+            This maps Kafka offsets just by converting to hexadecimals.
```
Comment (bocytko) [inline api/nakadi-event-bus-api.yaml]: Can you clarify that hex thing with an example?

Comment (rcillo) [inline api/nakadi-event-bus-api.yaml]: @bocytko example added bellow.

**Comment by tlossen** [comment]
minor nitpick: max integer if 7FFFFFFF (= 8 hex chars), max long is 7FFFFFFFFFFFFFFF (= 16 hex chars).

File: api/nakadi-event-bus-api.yaml [hunk 5]
```diff
@@ -1184,8 +1184,20 @@ definitions:
...
+          - offsets for empty partitions have a special value of BEGIN.
+          - offsets are fully ordered and comparable lexicographicaly.
+          - it's not possible to determine how many events there are between two offsets. Check
+            SubscriptionEventTypeStats for that.
+          - timeline and index are zero padded.
+        example: "0000000AB/0123456789ABCDEF"
```
Comment (bocytko) [inline api/nakadi-event-bus-api.yaml]: The length is now 26 (1 less than before).

File: api/nakadi-event-bus-api.yaml (hunk 4 above)
Comment (bocytko) [inline api/nakadi-event-bus-api.yaml]: The length is now 26 (1 less than before).

Comment (rcillo) [inline api/nakadi-event-bus-api.yaml]: Bad `wc` usage options. My mistake. Fixed.

**Comment by bocytko** [comment]
> we've seem some clients doing arithmetics with offsets. Something like `number_of_events_to_be_consumed = ...`

This is a natural concequence of exposing numeric cursors.

If clients do it now, why not provide the data they need directly within the API?
- number of available events since `last_offset`
- cursor for `last_offset`
- cursor for `next_offset`

Clients can use different operations:
a) Give me N events since `last_offset`
b) Give me the offset `last_offset + N`
c) Give me up to N events beginning with `next_offset` (long-poll).

The offsets could then be non-numeric strings like `uuids`, similar to database cursors where you have special functions for manipulating and fetching data from.

**Comment by tlossen** [comment]
@rcillo ok, let me rephrase my proposal. do we really need 92 bit for a simple offset? or can we maybe fit it into 64 bit, as described above? 

my main concern is ease of use / overhead when passing it around or storing it, as the largest base types in many systems or languages are 64 bit (for example long in java, bigint in postgres).

**Comment by jhorstmann** [comment]
Do we need to define this format at all in this detail? My only usecase as a library author is described at https://github.com/zalando-incubator/fahrschein/blob/0.4.4/src/main/java/org/zalando/fahrschein/CursorManager.java#L50 and any format that can be compared lexicographically (maybe after padding to the same length) would support that usecase. What are the usecases that other libraries or users have?

**Comment by rcillo** [comment]
After @tkrop and @bocytko suggestions, we are inclined to ditch this proposal in favor of a completely opaque cursor, which would contain only encoded data.

It would make room for further extensions at the cost of opaqueness. On the other hand, we would be adding endpoints for cursor manipulation, filling the gap left by opaque cursors.

**Comment by tlossen** [comment]
@rcillo in that case, it would be good if nakadi could provide something that represents the (per-partition) order of the stream, like the SequenceNumber in kinesis.

**Comment by rcillo** [comment]
We are closing this PR after collecting all your feedback. We'll be working on another proposal for opaque cursors.

@tlossen when working on this new proposal we'll study the viability of adding this SequenceNumber 👍
